src/bootstrapper.cc

```diff
@@ -51,6 +61,23 @@ void SetupNextTick(const FunctionCallbackInfo<Value>& args) {
   args.GetReturnValue().Set(ret);
 }
 
+void WeakCallback(
+    const v8::WeakCallbackInfo<Persistent<Promise> >& data) {
+  Environment* env = Environment::GetCurrent(data.GetIsolate());
+  Persistent<Promise>* p = data.GetParameter();
+  p->ClearWeak();
+
+  env->SetImmediate([](Environment* env, void* data) {
+    HandleScope handle_scope(env->isolate());
+    Persistent<Promise>* p = static_cast<Persistent<Promise>*>(data);
+    Local<Promise> promise = p->Get(env->isolate());
+    Local<Value> err = promise->Result();
+    Local<Message> message = Exception::CreateMessage(env->isolate(), err);
+    InternalFatalException(env->isolate(), err, message, kFromPromise);
```

Comment:
The documentation in this PR mentions:

> If set to `ERROR_ON_GC` unhandled rejections that are garbage collected are raised as uncaught exceptions.

However I don't think calling `InternalFatalException` is throwing any error. It's emitting an `uncaughtException` event though. 

If the intention is not to throw any error but just emit `uncaughtException`, I think it'd be useful to clarify the wording and mention that "any unhandled rejection that is garbage collected will emit an `uncaughtException` event".

Reply:
I think “throws an exception” is both correct and clearer. The uncaughtException event is caused by the thrown error, it’s never really emitted on its own.

Reply:
> I think “throws an exception” is both correct and clearer. 

I don't think it is what the current implementation does though. It does not seem to throw an exception. It seems to call `InternalFatalException`, which itself (and its callees) does not throw an exception. 

> The uncaughtException event is caused by the thrown error, it’s never really emitted on its own.

I'm familiar with when the `uncaughtException` event is generally emitted. In the current implementation of this PR though I think it is emitted directly by calling `InternalFatalException` which calls `process._fatalException`, and not by actually throwing an error.

I could be missing something though.

Reply:
You might be right, for sure - but that’s not observable to the user, and it’s not necessary to document the implementation details. I think sticking with the paradigm of “throwing an exception” is clearer for documentation.

Reply:
It is observable by users of the `--abort-on-uncaught-exception` command line option. When using this command line option with the `NODE_UNHANDLED_REJECTION=ERROR_ON_GC` mode, and when rejecting a promise without an error handler, the process exits but does not abort, which seems unexpected.

Reply:
right, but conceptually (to me) that's aborting on an uncaught exception - not aborting on the `uncaughtException` event being emitted.

Reply:
@ljharb what Julien means here is that people who run Node with `--abort-on-uncaught-exception` do so for post-mortem debugging and expect to have a core dump to look at when the process exits.

Reply:
@benjamingr that certainly seems reasonable to me - that sounds like you're saying that the implementation needs to legitimately throw an exception, as opposed to making the documentation say it emits an event?

Reply:
Well I don't feel strongly about the implementation actually throwing an exception. 

I feel like the implementation should interop with other Node.js features like `--abort-on-uncaught-exception` - whether or not an actual exception is thrown isn't really important to me but I'd like users to be able to look at a core dump. See https://github.com/nodejs/node/pull/20097#discussion_r196619398

Reply:
@ljharb sorry for being confusing then :) The implementation can choose to do whatever it wants so that with `--abort-on-uncaught-exception` and `ERROR_ON_GC` or `ERROR` it will allow the same functionality that exceptions do.

That is - a core dump file is generated for post mortem analysis.

I'm also fine with that being a TODO at this point - I want to seek consensus on that being the correct behaviour.